core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -160,7 +160,10 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
       override def beforeInitializingSession(): Unit = {
         val expireEvent = new Expire
         eventManager.clearAndPut(expireEvent)
-        expireEvent.waitUntilProcessed()
+
+        // Block initialization of the new session until the expiration event is being handled to
+        // ensure it cannot be associated with the wrong session
```

Comment:
This is mainly to ensure that all pending events in the queue are processed before we create a new ZK session.